core/util/src/main/java/org/eclipse/rdf4j/common/iteration/ConvertingIteration.java

```diff
@@ -56,10 +58,18 @@ protected abstract T convert(S sourceObject)
 	 * @return <tt>true</tt> if the source type iteration contains more elements, <tt>false</tt> otherwise.
 	 * @throws X
 	 */
+	@Override
 	public boolean hasNext()
 		throws X
 	{
-		return iter.hasNext();
+		if (isClosed()) {
+			return false;
+		}
+		boolean result = iter.hasNext();
+		if (!result) {
+			close();
```

Comment:
I seem to remember that I proposed something very similar earlier (in Sesame) and there were some negative responses to this approach, so we abandoned it. See https://bitbucket.org/openrdf/sesame/pull-requests/550/ses-2323-added-autoclose-of-exhausted/diff